examples/integer-localizations.jl

```diff
@@ -0,0 +1,222 @@
+    # Perform some sanity checks
+    b in inverted_set(R) || error("$b does not belong to the units of $R")
+    return new{MultSetType}(a, b, R)
+  end
+end
+
+### required getter functions
+parent(f::FmpzLocalizedRingElem{MultSetType}) where {MultSetType} = f.R::FmpzLocalizedRing{MultSetType}
+numerator(f::FmpzLocalizedRingElem{MultSetType}) where {MultSetType} = f.numerator::fmpz
+denominator(f::FmpzLocalizedRingElem{MultSetType}) where {MultSetType} = f.denominator::fmpz
+
+### required implementation of the arithmetic
+function Base.:(//)(a::Oscar.IntegerUnion, b::FmpzLocalizedRingElem)
+  c = ZZ(a)
+  g = gcd(c, numerator(b))
+  c = divexact(c, g)
+  numerator(b) in inverted_set(parent(b)) || error("the second argument is not a unit in this local ring")
+  return reduce_fraction((parent(b))(c*denominator(b), divexact(numerator(b), g)))
+end
+
+function Base.:(//)(a::T, b::T) where {T<:FmpzLocalizedRingElem}
+  parent(a) == parent(b) || error("the arguments do not have the same parent ring")
+  g = gcd(numerator(a), numerator(b))
+  c = divexact(numerator(a), g)
+  d = divexact(numerator(b), g)
+  d in inverted_set(parent(b)) || error("the second argument is not a unit in this local ring")
+  return reduce_fraction((parent(a))(c*denominator(b), d*denominator(a)))
+end
+
+### optional enhancement of the arithmetic
+function reduce_fraction(f::FmpzLocalizedRingElem{MultSetType}) where {MultSetType}
+  g = gcd(numerator(f), denominator(f))
+  return FmpzLocalizedRingElem(parent(f), divexact(numerator(f), g), divexact(denominator(f), g))
+end
+
+### required conversions
+(W::FmpzLocalizedRing{MultSetType})(f::fmpz) where {MultSetType} = FmpzLocalizedRingElem(W, f, one(ZZ))
+(W::FmpzLocalizedRing{MultSetType})(a::fmpz, b::fmpz) where {MultSetType} = FmpzLocalizedRingElem(W, a, b)
+
+### additional conversions
+(W::FmpzLocalizedRing{MultSetType})(q::fmpq) where {MultSetType} = FmpzLocalizedRingElem(W, numerator(q), denominator(q))
+(W::FmpzLocalizedRing{MultSetType})(q::Rational{Int}) where {MultSetType} = FmpzLocalizedRingElem(W, ZZ(numerator(q)), ZZ(denominator(q)))
+@Markdown.doc """
+    (W::FmpzLocalizedRing{MultSetType})(i::Oscar.IntegerUnion) where {MultSetType} = FmpzLocalizedRingElem(W, ZZ(i), one(ZZ))
+
+Part of the minimal interface for localized rings. This routine returns the conversion of 
+an integer i to an element i//1 ∈ W.
+"""
+(W::FmpzLocalizedRing{MultSetType})(i::Oscar.IntegerUnion) where {MultSetType} = FmpzLocalizedRingElem(W, ZZ(i), one(ZZ))
+(W::FmpzLocalizedRing{MultSetType})(a::Oscar.IntegerUnion, b::Oscar.IntegerUnion) where {MultSetType} = FmpzLocalizedRingElem(W, ZZ(a), ZZ(b))
+
+### implementation of Oscar's general ring interface
+one(W::FmpzLocalizedRing{MultSetType}) where {MultSetType} = W(1)
+zero(W::FmpzLocalizedRing{MultSetType}) where {MultSetType} = W(0)
+
+elem_type(W::FmpzLocalizedRing{MultSetType}) where {MultSetType} = FmpzLocalizedRingElem{MultSetType}
```

Comment:
this file contains *tons* of redundant `where` clauses (and there are more elsewhere, too)